src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandle.cs

```diff
@@ -0,0 +1,85 @@
+    /// with LocalAlloc for Marshal.AllocHGlobal so it doesn't hit the super slow path.
+    /// 
+    /// Windows attempts to grab space from the low fragmentation heap if the requested memory is below a platform specific
+    /// threshold and certain flags aren't in play (such as NO_SERIALIZE).
+    /// </remarks>
+    internal sealed class SafeHeapHandle : SafeBuffer
+    {
+        /// <summary>
+        /// The handle for the process heap.
+        /// </summary>
+        private static readonly IntPtr s_processHeap = Interop.mincore.GetProcessHeap();
+
+        /// <summary>
+        /// Allocate a buffer of the given size and zero memory if requested.
+        /// </summary>
+        /// <param name="byteLength">Required size in bytes. Must be less than UInt32.MaxValue for 32 bit or UInt64.MaxValue for 64 bit.</param>
+        /// <exception cref="OutOfMemoryException">Thrown if the requested memory size cannot be allocated.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if size is greater than the maximum memory size.</exception>
+        public SafeHeapHandle(ulong byteLength, bool zeroMemory = false) : base(ownsHandle: true)
+        {
+            Resize(byteLength, zeroMemory);
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        /// <summary>
+        /// Resize the buffer to the given size and zero memory if requested.
+        /// </summary>
+        /// <param name="byteLength">Required size in bytes. Must be less than UInt32.MaxValue for 32 bit or UInt64.MaxValue for 64 bit.</param>
+        /// <exception cref="OutOfMemoryException">Thrown if the requested memory size cannot be allocated.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if size is greater than the maximum memory size.</exception>
+        public void Resize(ulong byteLength, bool zeroMemory = false)
+        {
+            if (IsClosed) throw new ObjectDisposedException("HeapHandle");
+
+            uint flags = zeroMemory ? Interop.mincore.HEAP_ZERO_MEMORY : 0;
+
+            if (handle == IntPtr.Zero)
+            {
+                handle = Interop.mincore.HeapAlloc(s_processHeap, flags, (UIntPtr)byteLength);
+            }
+            else
+            {
+                // This may or may not be the same handle, Windows may realloc in place. If the
+                // handle changes Windows will deal with the old handle, trying to free it will
+                // cause an error.
+                handle = Interop.mincore.HeapReAlloc(s_processHeap, flags, handle, (UIntPtr)byteLength);
+            }
+
+            if (handle == IntPtr.Zero)
+            {
+                // Only real plausible answer
+                throw new OutOfMemoryException();
+            }
+
+            Initialize(byteLength);
```

Comment:
Should we be using GC.AddMemoryPressure/RemoveMemoryPressure so that the GC is aware of these potentially large allocations being made not under its control?  Otherwise it won't know that collection could potentially be valuable.